### ISSUE TITLE
SVT-HEVC multi-instance are blocked when calling deinit by multi-thread

### ISSUE THREAD

**Issue by Yeongjin-Jeong** [OpenVisualCloud/SVT-HEVC#357]
I used the gstreamer svthevcenc plugin, and found that if I create a multi-instance by allocating 1 instance per 1 thead, sometimes it would blocked when calling deinit per thread. So they are not terminated. This occurred when cpu utilization was very high.

Does SVT-HEVC support multi-instance by multi-thread in single process? I have checked that multi-instance works fine by single thread.

The command used is:
```
gst-launch-1.0 videotestsrc num-buffers=1800 ! video/x-raw,width=1920,height=1080,framerate=30/1 ! queue ! tee name=t \
t. ! queue ! svthevcenc speed=9 key-int-max=30 bitrate=6000 ! h265parse ! queue ! fakesink \
t. ! queue ! svthevcenc speed=9 key-int-max=30 bitrate=6000 ! h265parse ! queue ! fakesink \
... (branches for high cpu utilization)
```

**Comment by tianjunwork** [comment]
Hi @Yeongjin-Jeong , just to confirm, what you mean "multi-instance" is "The multi-instance functionality" mentioned in the README or the way you use the plugin?
The multi-instance functionality is an application level feature, one SvtHevcEncApp can launch multiple encoder instances by using -nch N.

**Comment by Yeongjin-Jeong** [comment]
That "multi-instance" means multiple svthevcenc elements. I used gstreamer for creating multiple outputs of the same input in the same process. Each svthevcenc element uses libSvtHevcEnc through its own thread.

**Comment by Yeongjin-Jeong** [comment]
@gokul989, If you want to run on single thread, don't use the queue element before the svthevcenc element.

Such as:
* single process with multi thread
```
gst-launch-1.0 videotestsrc num-buffers=1800 ! video/x-raw,width=1920,height=1080,framerate=30/1 ! queue ! tee name=t \
t. ! queue ! svthevcenc speed=9 key-int-max=30 bitrate=6000 ! h265parse ! queue ! fakesink \
t. ! queue ! svthevcenc speed=9 key-int-max=30 bitrate=6000 ! h265parse ! queue ! fakesink \
... (many branches for high cpu utilization)
```
* single process with single thread
```
gst-launch-1.0 videotestsrc num-buffers=1800 ! video/x-raw,width=1920,height=1080,framerate=30/1 ! queue ! tee name=t \
t. ! svthevcenc speed=9 key-int-max=30 bitrate=6000 ! h265parse ! queue ! fakesink \
t. ! svthevcenc speed=9 key-int-max=30 bitrate=6000 ! h265parse ! queue ! fakesink \
... (many branches for high cpu utilization)
```

**Comment by gokul989** [comment]
Hi @Yeongjin-Jeong . Thanks for more clarification. I was told that support for multi-instancing feature on multi threads needs to be implemented.

**Comment by tianjunwork** [comment]
Hi @Yeongjin-Jeong, Sai @gokul989 tried today your command by ending the execution by ctrl + c. He didn't see deinit being called, so no blocking. Could you let us know how to trigger the blocking?

**Comment by tianjunwork** [comment]
Hi @Yeongjin-Jeong , the root cause is the use of global memoryMap in EbEncHandle.c which is not thread safe. 
In gst case, each svthevcenc instance is running in its own thread, memoryMap of each encoder instance is overridden by the last one from call of InitH265EncoderHandle. So all the resources are recorded in only one memoryMap.

When first EbDeinitEncoder is being called, EbDestroyThread is called on all the threads recorded in memoryMap.
When second EbDeinitEncoder is being called, the thread IDs in memoryMap are already invalid. So thread_join() never return.

Good news is that AV1 already have a fix in:  https://github.com/OpenVisualCloud/SVT-AV1/pull/314.
I will port it over to SVT-HEVC. Since it is not a quick change, it will take some time.

**Comment by tianjunwork** [comment]
Hi @Yeongjin-Jeong , Austin had PR #446, as a quick fix for this issue.
1. Please check whether the issue could be addressed. We verified that N instances could work with gstreamer;
2. As long term solution, we will port AV1 PR #314 to keep the code base sync.

**Comment by tianjunwork** [comment]
This issue is fixed by #467. Verified with command line provided by @Yeongjin-Jeong and more threads.
